### ISSUE TITLE
[5.0] DB transactions don't work in tests (same way as in L4)

### ISSUE THREAD

**Issue by neomerx** [laravel/framework#7017]
Easiest way to reproduce
 - configure correct .env file
 - run the following test
~~~
<?php

class ExampleTest extends TestCase
{
    public function setUp()
    {
        parent::setUp();
        echo PHP_EOL.'Before opening transaction... ';
        DB::beginTransaction();
        echo 'Transaction opened'.PHP_EOL;
    }

    public function tearDown()
    {
        parent::tearDown();
        echo 'Before closing transaction... ';
        DB::rollBack(); // <-- fails here
        echo 'Transaction closed'.PHP_EOL;
    }

	public function testBasicExample()
	{
	}
}
~~~
Output:
Before opening transaction... Transaction opened
Before closing transaction... 

Time: 84 ms, Memory: 13.00Mb

There was 1 error:

1) ExampleTest::testBasicExample
ReflectionException: Class config does not exist

What I found is that before rollBack() Illuminate\Container\Container->instances had an array of various instances but on rollBack() it was empty which caused
~~~
Cannot use object of type Illuminate\Support\Facades\Config as array in ...../vendor/laravel/framework/src/Illuminate/Database/DatabaseManager.php on line 259

**Comment by neomerx** [comment]
NB It works if parent::tearDown(); moved down the test howerver this code worked fine in L4

**Comment by GrahamCampbell** [comment]
This is actually totally unrelated. This is caused by the ioc container not being initialised properly by your test suite.

**Comment by neomerx** [comment]
@GrahamCampbell How it should be initialized? I've been porting real tests from L4 to L5 and proposed 'fix' works
~~~
//        parent::tearDown(); // <-- if uncommented instead of the last line tests will fail
        DB::rollBack();
        DB::disconnect();
        Mockery::close();
        parent::tearDown();

**Comment by GrahamCampbell** [comment]
It looks more like you've got some static stuff left over from a previous test.

**Comment by neomerx** [comment]
@GrahamCampbell I took fresh L5 and added only 1 test. So I didn't have any my stuff at all.
As for real code I've ported about 30 tests and moving parent::tearDown() to the last line solved my problems.
It would be nice if Container thrown a meaningful exception rather than fail with type incompatibility exception.

**Comment by GrahamCampbell** [comment]
The container can't know that what's its returning isn't what you want. It's not that magic I'm afraid.

**Comment by neomerx** [comment]
@GrahamCampbell Sure you know the code better :) I've debugged it a bit and found suspicious when container worked fine for giving 'config' many times and then it become empty before test really finished.

**Comment by jdgriffith** [comment]
I found on my L5 instance this was happening due to the way the config file gets loaded in the DatabaseManager in tests.

For instance: 

`$connection->setFetchMode($this->app['config']['database.fetch']);`

Instead of:

`$connection->setFetchMode(\Config::get('database.fetch'));`

Not sure if I am just initializing it wrong as well. But when I change it to this everything runs fine. Tests   and the application.

Should this be reopened or moved to a new ticket?

**Comment by GrahamCampbell** [comment]
What? Both statements you have there are identical under the hood, only the first one is more correct.

**Comment by jdgriffith** [comment]
It may be "more correct" but it doesn't work with my tests. And changing it does not break my application either from the limited checking I did.

**Comment by jdgriffith** [comment]
I should mention I am using Codeception for some tests.

**Comment by GrahamCampbell** [comment]
You're probably doing the tests wrong then. Ask on the forums.

**Comment by jdgriffith** [comment]
So just to verify the fix for this situation (according to this issue) is just move parent::tearDown() to the last line in most cases?

**Comment by mmochetti** [comment]
Hi! I'm having the same issue with Lumen. How did you solve it? thanks

**Comment by neomerx** [comment]
Haven't tried transactions in Lumen tests yet

**Comment by hengsoheak** [comment]
Laravel 5.3.16

Database Transaction  is really not work

   if ($this->request->isMethod('post')) {

            try {

                DB::beginTransaction();
                $cats = new Cat();
                $res = [];
                $cats->parent_id = $this->request->input('category_id');
                $cats->status = ($this->request->input('status')) ? $this->request->input('status') : 0;
                if ($res['cat'] = $cats->save()) {

                    $catD = new CategoryDescriptions();
                    $catD->category_id = $cats->id;
                    $catD->language_id = 1;
                    $catD->name = $this->request->input('en_name');
                    if ($res['cat2'] = $catD->save()) {
                        $catD2 = new CategoryDescriptions();
                        $catD2->category_id = $cats->id;
                        $catD2->language_id = 2;
                        $catD2->name = $this->request->input('kh_name');
                        $res['all'] = $catD2->save();
                    }
                }
                if(!empty($res['all'])) {
                    //DB::commit();
                    return $res;
                }
            } catch (Exception $e) {
                DB::rollback();
            }
            return [false];
        }

**Comment by CalvinScott** [comment]
After literal days spent trying to figure out why some data being created in one test was being persisted through to another test (tried using both DatabaseTransactions and RefreshDatabase), I finally got the tests/database to actually reflect what I wanted by following OP's suggestion. 

I think the folks at @laravel-internals/@laracasts seriously need to reflect on their DatabaseTransactions and RefreshDatabase code and documentation. Don't just shrug it off as "this is unrelated" or some other BS, as clearly there is something going on here that's actually affecting people. The documentation around those traits is so nebulous it's almost funny. "The RefreshDatabase trait takes the most optimal approach" ...ok, and how? "Use the trait on your test class and everything will be handled for you" ...ok, it doesn't? Not to mention the example code is just the ```use XXX``` at the top of the class, which is far from helpful when it _doesn't actually handle it for you_.

Just for some background on what I wasted my time on, all I wanted was DatabaseTransactions to _work_, so I could run migrations once per test file, and seed some custom data *once* for each test file. This isn't possible using `DatabaseMigrations`, as after each migration you would need to re-seed any test data, which can get extremely expensive if your Test suite needs a lot of test data and/or you have a lot of tests.

Everything worked with DatabaseMigrations (which makes sense as it basically drops all tables before starting fresh), but when I simply replaced with DatabaseTransactions/RefreshDatabase, everything went to shit. Some tests would pass, while others wouldn't. Sometimes data persisted between tests, while other times it wouldn't. Connection was never a problem, even though this is the oft-diagnosed problem, and even when using a non-default MySQL connection. 

Like I said, I've finally gotten the tests/database to behave in a sensible way by skipping all of the provided Laravel traits, and using the DB::beginTransaction and DB::rollBack methods within the setUp and tearDown functions, along with single-run custom initTest() function that only gets run on the first setUp so that we can only run the seeders once. This was all done on a basically fresh Laravel app
